dev/TabView/TabView.cpp

```diff
@@ -0,0 +1,260 @@
+                        m_isTabClosing = true;
+                        m_indexToSelect = index;
+                        break;
+                    }
+
+                    // try the next item
+                    index++;
+                    if (index >= (int)Items().Size())
+                    {
+                        index = 0;
+                    }
+                } while (index != startIndex);
+            }
+        }
+    }
+
+    UpdateTabWidths();
+
+    __super::OnItemsChanged(item);
+}
+
+void TabView::OnSelectionChanged(const winrt::IInspectable& sender, const winrt::SelectionChangedEventArgs& args)
+{
+    if (m_isTabClosing)
+    {
+        m_isTabClosing = false;
+        SelectedItem(Items().GetAt(m_indexToSelect));
+    }
+
+    UpdateTabContent();
+}
+
+void TabView::UpdateTabContent()
+{
+    if (auto tabContentPresenter = m_tabContentPresenter.get())
+    {
+        if (!SelectedItem())
+        {
+            tabContentPresenter.Content(nullptr);
+            tabContentPresenter.ContentTemplate(nullptr);
+        }
+        else
+        {
+            auto container = ContainerFromItem(SelectedItem()).as<winrt::ListViewItem>();
+            if (container)
+            {
+                tabContentPresenter.Content(container.Content());
+                tabContentPresenter.ContentTemplate(container.ContentTemplate());
+            }
+        }
+    }
+}
+
+void TabView::CloseTab(winrt::TabViewItem container)
+{
+    if (auto item = ItemFromContainer(container))
+    {
+        uint32_t index = 0;
+        if (Items().IndexOf(item, index))
+        {
+            auto args = winrt::make_self<TabViewTabClosingEventArgs>(item);
+
+            m_tabClosingEventSource(*this, *args);
+
+            if (!args->Cancel())
+            {
+                Items().RemoveAt(index);
+            }
+        }
+    }
+}
+
+void TabView::OnScrollDecreaseClick(const winrt::IInspectable& sender, const winrt::RoutedEventArgs& args)
+{
+    if (auto scrollViewer = m_scrollViewer.get())
+    {
+        scrollViewer.ChangeView(std::max(0.0, scrollViewer.HorizontalOffset() - c_scrollAmount), nullptr, nullptr);
+    }
+}
+
+void TabView::OnScrollIncreaseClick(const winrt::IInspectable& sender, const winrt::RoutedEventArgs& args)
+{
+    if (auto scrollViewer = m_scrollViewer.get())
+    {
+        scrollViewer.ChangeView(std::min(scrollViewer.ScrollableWidth(), scrollViewer.HorizontalOffset() + c_scrollAmount), nullptr, nullptr);
+    }
+}
+
+void TabView::UpdateTabWidths()
+{
+    if (TabWidthMode() == winrt::TabViewWidthMode::SizeToContent)
+    {
+        for (int i = 0; i < (int)(Items().Size()); i++)
+        {
+            auto container = ContainerFromItem(Items().GetAt(i)).as<winrt::ListViewItem>();
+            if (container)
+            {
+                container.Width(DoubleUtil::NaN);
+            }
+        }
+    }
+    else
+    {
```

Comment:
Normally I'd expect this kind of thing in VSM setters, what was the motivation for doing it from code?

Reply:
Because I have to do math...?

Reply:
And is the TemplateSettings pattern too annoying/tedious?